docs/rules/no-access-state-in-setstate.md

```diff
@@ -0,0 +1,39 @@
+# Prevent using this.state within a this.setState (no-access-state-in-setstate)
+
+This rule should prevent usage of this.state inside setState calls.
+Such usage of this.state might result in errors when two state calls is
+called in batch and thus referencing old state and not the current
+state. An example can be an increment function:
+
+```
+function increment() {
+  this.setState({value: this.state.value + 1});
+}
+```
+
+If these two setState operations is grouped together in a batch it will
+look be something like the following, given that value is 1:
+
+```
+setState({value: 1 + 1})
+setState({value: 1 + 1})
+```
+
+This can be avoided with using callbacks which takes the previous state
+as first argument:
+
+```
+function increment() {
+  this.setState(prevState => ({value: prevState.value + 1}));
```

Comment:
This is good advice when the semantic is "increment", but referencing this.state inside a setState call doesn't necessarily mean that the dev doesn't want s "snapshot".

What happens with:
```js
const { value } = this.state;
this.setState({ value: value + 1 });
```
?

This is conceptually the same code, but it might be a useful way for the author to indicate that this is an intentional ordering.